### PR TITLE
doc: add shared buffer doc

### PR DESCRIPTION
**Pull request by kwannoel**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Not sure who is the original author, the quip doc doesn't write it. I assume it is @hzxa21. I migrated it from those docs on shared buffer in quip.

I have done some light edits to refer to the shared buffer in a present tense, and to clarify some parts.
For reviewers, please leave comments on outdated parts if any.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: docs/shared-buffer.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+- [Introduction](#introduction)
+- [Part 1: Async Checkpoint](#part-1-async-checkpoint)
+  - [Write Path](#write-path)
+  - [Read Path](#read-path)
+- [Part 2: Write Anytime / Async Flush](#part-2-write-anytime--async-flush)
+  - [A New Merge Iterator](#a-new-merge-iterator)
+    - [Considerations](#considerations)
+  - [TopN Executor (Merge-On-Read)](#example-topn-executor-merge-on-read)
+  - [HashJoinExecutor](#the-new-hashjoinexecutor)
+  - [The New ExtremeState](#the-new-extremestate)
```
Comment (fuyufjh) [inline docs/shared-buffer.md:14]: These 3 sections are missing.

Comment (kwannoel) [reply docs/shared-buffer.md]: These are outdated examples. Removed.

File: docs/shared-buffer.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+- Batch writes on worker node level, so as to reduce SST number.
+
+  - Currently, a single epoch might produce hundreds of SST, which makes meta service hard to handle.
+
+- Support async checkpoint.
+
+  - The shared buffer will generate SST based on epoch, and provide a consistent view of a epoch, by merging the snapshot of the storage SSTs and the immutable in-memory buffers.
+
+- Support read-after-write, so as to make executor logic simpler (so-called async flush).
```
[the remark this answers is not in the document]
Comment (kwannoel) [reply docs/shared-buffer.md]: Yes, restructured this sentence to make that part clear.

File: docs/shared-buffer.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+The async checkpoint logic will be implemented with StreamManager (on meta) and HummockLocalManager (on CN), and we will only have immutable memtables in this part.
+
+### Write Path
+
+Assume there is a new barrier flowing across our system.
+
+When each executor consumes a barrier, it will produce a write batch (which should complete immediately) and forward the barrier.
+
+After all executors have completed processing barrier, a checkpoint will be initiated, and new SST of the previous epoch will be produced. As our system only have one barrier flowing, we can simply checkpoint epoch 3 after stream manager collects barrier 4. In the future, we will need to record the dependencies of barriers (e.g., barrier 3 is immediately followed by barrier 4), and decide which epoch to checkpoint when new barriers have been collected.
```
[the remark this answers is not in the document]
Comment (kwannoel) [reply docs/shared-buffer.md:71]: Yeah I don't see the benefit either. @hzxa21 any comments for this part? Removing it first.

[the remark this answers is not in the document]
Comment (kwannoel) [reply docs/shared-buffer.md]: 5baf90889f89b1eb5ca8adf8aac304f9e2563156

**Review by wenym1 — COMMENTED** [review]
The quip doc seems out of time. There is many new designs after the doc. 

The PR can be merged first if it's just for doc migration. We may consider adding more up-to-date details to the doc later. cc @hzxa21

**Comment by kwannoel** [comment]
> The quip doc seems out of time. There is many new designs after the doc.
> 
> The PR can be merged first if it's just for doc migration. We may consider adding more up-to-date details to the doc later. cc @hzxa21

Yes, that will be much appreciated.

**Review by wenym1 — APPROVED** [review]
Let's merge it first. Can you create a tracking issue on updating the doc of shared buffer? @kwannoel

**Pull request merged**